### PR TITLE
Add support for  show(penum char |....|)

### PR DESCRIPTION
**Pull request by mo-xiaoming**

### CODE REVIEW

File: lib/hobbes/boot/show.hob [hunk 1]
```diff
@@ -119,6 +119,8 @@ instance (Ord t t) => Ord (penum t v) (penum t v) where
   x >= y = x.t >= y.t
 instance (UnpackEnum t v, Show v) => Show t where
   show e = show(unpackEnum(e))
+instance Show ((penum a v)) where
+  show = penumShow
```
Comment (smunix) [inline lib/hobbes/boot/show.hob:123]: nice, thanks @mo-xiaoming

File: lib/hobbes/read/pgen/hexpr.y [hunk 2]
```diff
@@ -229,6 +230,22 @@ MonoTypePtr makeRecType(const Record::Members& tms) {
   return Record::make(tms);
 }
 
+MonoTypePtr makePVarType(const Variant::Members &vms,
```
Comment (smunix) [inline lib/hobbes/read/pgen/hexpr.y:233]: Let's change `Variant::make(vms)` to use `makePVarType(...)` [in the C++ constructor for penum](https://github.com/morganstanley/hobbes/blob/90c9865cfbfb3c4c08437dcd27bd9a090028ff99/include/hobbes/lang/tylift.H#L413) too

**Review by smunix — CHANGES_REQUESTED** [review]

File: lib/hobbes/read/pgen/hexpr.y (hunk 2 above)
Comment (mo-xiaoming) [reply lib/hobbes/read/pgen/hexpr.y]: Should this check be added to `Variant::make` itself, just wondering?

**Comment by mo-xiaoming** [comment]
Should `|A|::((penum char |A('A'), B('B')|))` be supported? currently, it is a syntax error

**Comment by smunix** [comment]
> Should `|A|::((penum char |A('A'), B('B')|))` be supported? currently, it is a syntax error

Would be nice to have indeed

File: lib/hobbes/read/pgen/hexpr.y [hunk 3]
```diff
@@ -993,6 +1028,17 @@ mvarlist: mvarlist "," id ":" l0mtype { $$ = $1;
         | id ":" l0mtype              { $$ = autorelease(new Variant::Members()); $$->push_back(Variant::Member(*$1, *$3,                0)); }
         | id                          { $$ = autorelease(new Variant::Members()); $$->push_back(Variant::Member(*$1, Prim::make("unit"), 0)); }
 
+mpvarlist: mpvarlist "," id "(" "intV" ")"   { $$ = $1;                                  $$->push_back(Variant::Member(*$3, Prim::make("unit"), $5)); }
```
Comment (smunix) [inline lib/hobbes/read/pgen/hexpr.y:1031]: let's simplify this?
```yacc
mpvarlist = mpvarlist "," mpvar  
               | mpvar
mpvar = id "(" intV ")" 
            | id "(" charV ")" 
            | id "(" boolV ")" 
            | ...
```

**Review by smunix — CHANGES_REQUESTED** [review]

**Review by smunix — APPROVED** [review]

**Pull request merged**